url696d9a93c9_liaosiwei_bayesnetwork_tgz-pJ8-tutorial_TutorialJ8/src/com/alibaba/Train.java

```diff
@@ -0,0 +1,352 @@
+            intersection.retainAll(notSaveSet);
+            intersection.retainAll(notCarSet);
+            beforeProbability = intersection.size();
+            if (beforeProbability == 0) {
+                buyProb[6] = 0.5;
+                buyProb[7] = 0.5;
+            }
+            else {
+                intersection.retainAll(buySet);
+                unionProb = intersection.size();
+                buyProb[6] = unionProb / beforeProbability;
+                buyProb[7] = 1 - buyProb[6];
+            }
+
+            intersection = new HashSet<String>(notClickSet);
+            intersection.retainAll(saveSet);
+            intersection.retainAll(carSet);
+            beforeProbability = intersection.size();
+            if (beforeProbability == 0) {
+                buyProb[8] = 0.6;
+                buyProb[9] = 0.4;
+            }
+            else {
+                intersection.retainAll(buySet);
+                unionProb = intersection.size();
+                buyProb[8] = unionProb / beforeProbability;
+                buyProb[9] = 1 - buyProb[8];
+            }
+
+            intersection = new HashSet<String>(notClickSet);
+            intersection.retainAll(saveSet);
+            intersection.retainAll(notCarSet);
+            beforeProbability = intersection.size();
+            if (beforeProbability == 0) {
+                buyProb[10] = 0.5;
+                buyProb[11] = 0.5;
+            }
+            else {
+                intersection.retainAll(buySet);
+                unionProb = intersection.size();
+                buyProb[10] = unionProb / beforeProbability;
+                buyProb[11] = 1 - buyProb[10];
+            }
+
+            intersection = new HashSet<String>(notClickSet);
+            intersection.retainAll(notSaveSet);
+            intersection.retainAll(carSet);
+            beforeProbability = intersection.size();
+            if (beforeProbability == 0) {
+                buyProb[12] = 0.5;
+                buyProb[13] = 0.5;
+            }
+            else {
+                intersection.retainAll(buySet);
+                unionProb = intersection.size();
+                buyProb[12] = unionProb / beforeProbability;
+                buyProb[13] = 1 - buyProb[12];
+            }
+
+            intersection = new HashSet<String>(notClickSet);
+            intersection.retainAll(notSaveSet);
+            intersection.retainAll(notCarSet);
+            beforeProbability = intersection.size();
+            if (beforeProbability == 0) {
+                buyProb[14] = 0.2;
+                buyProb[15] = 0.8;
+            }
+            else {
+                intersection.retainAll(buySet);
+                unionProb = intersection.size();
+                buyProb[14] = unionProb / beforeProbability;
+                buyProb[15] = 1 - buyProb[14];
+            }
+
+            initNetwork();
+            for (String key2: table.get(key1).keySet()) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.